File: Src/Graphology.Tests/DefaultGetTypeNameStringTests.cs
using Ploeh.AutoFixture.Idioms;
using Ploeh.AutoFixture.Xunit;
using System;
using TeaDriven.Graphology.Tests.Conventions;
using Xunit;
using Xunit.Extensions;

namespace TeaDriven.Graphology.Tests
{
    public class DefaultGetTypeNameStringTests
    {
        [Theory, AutoNSubstituteData]
        public void Constructor_HasNullGuards(GuardClauseAssertion assertion)
        {
            assertion.Verify(typeof(DefaultGetTypeNameString).GetConstructors());
        }

        [Theory, AutoNSubstituteData]
        public void Constructor_ReturnsIGetTypeNameString(DefaultGetTypeNameString sut)
        {
            Assert.IsAssignableFrom<IGetTypeNameString>(sut);
        }

        [Theory, AutoNSubstituteData]
        public void For_TypeHasNullGuard(DefaultGetTypeNameString sut)
        {
            // Arrange
            string @out;

            // Act
            // Assert
            Assert.Throws<ArgumentNullException>(() => sut.For(null, out @out));
        }

        [Theory, AutoNSubstituteData]
        public void For_ReturnsTrueAndTypeName(DefaultGetTypeNameString sut)
        {
            // Arrange
            var type = typeof(AutoNSubstituteDataAttribute);

            var expectedResult = true;
            var expectedTypeName = "AutoNSubstituteDataAttribute";

            // Act
            string resultTypeName;
            var result = sut.For(type, out resultTypeName);

            // Assert
            Assert.True(result);

            Assert.Equal(expectedTypeName, resultTypeName);
        }
    }
}

File: Src/Graphology.TestsF/DefaultGetTypeNameStringTests.fs
namespace TeaDriven.Graphology.Tests

open Xunit
open Xunit.Extensions
open Ploeh.AutoFixture
open Ploeh.AutoFixture.Idioms
open FsUnit.Xunit

module DefaultGetTypeNameStringTests =
    open System
    open TeaDriven.Graphology.Visualization

    [<Theory; AutoFoqData>]
    let ``Constructors should have null guards`` (assertion : GuardClauseAssertion) =
        assertion.Verify (typeof<DefaultGetTypeNameString>.GetConstructors())

    [<Theory; AutoFoqData>]
    let ``Class should implement IGetTypeNameString`` (sut : DefaultGetTypeNameString) =
        Assert.IsAssignableFrom<IGetTypeNameString> sut

    [<Theory; AutoFoqData>]
    let ``For() should have null guards`` (sut : DefaultGetTypeNameString) =
        Assert.Throws<ArgumentNullException>(fun () -> sut.For null |> ignore)

    [<Theory; AutoFoqData>]
    let ``For() returns true and correct name`` (sut : DefaultGetTypeNameString) =
        let inputType = typeof<FactAttribute>

        let expectedName = "FactAttribute"

        let result, resultName = sut.For inputType

        result |> should be True
        resultName |> should equal expectedName